smarts/env/gymnasium/utils/observation_conversion.py

```diff
@@ -0,0 +1,1050 @@
+_DISCRETE2_SPACE = gym.spaces.Discrete(n=2)
+
+
+class BaseSpaceFormat:
+    def format(self, obs: Observation):
+        raise NotImplementedError()
+
+    def active(self, agent_interface: AgentInterface) -> bool:
+        raise NotImplementedError()
+
+    @property
+    def name(self):
+        raise NotImplementedError()
+
+    @property
+    def space(self):
+        raise NotImplementedError()
+
+
+class HeadingSpaceFormat(BaseSpaceFormat):
+    @property
+    def name(self):
+        return "heading"
+
+    @property
+    def space(self):
+        return _SIGNED_RADIANS_FLOAT32_SPACE
+
+
+class SpeedSpaceFormat(BaseSpaceFormat):
+    @property
+    def name(self):
+        return "speed"
+
+    @property
+    def space(self):
+        return _UNSIGNED_FLOAT32_SPACE
+
+
+class PositionSpaceFormat(BaseSpaceFormat):
+    @property
+    def name(self):
+        return "position"
+
+    @property
+    def space(self):
+        return _VEC3_SIGNED_FLOAT64_SPACE
+
+
+class VelocitySpaceFormat(BaseSpaceFormat):
+    @property
+    def space(self):
+        return _VEC3_SIGNED_FLOAT32_SPACE
+
+
+class AccelerationSpaceFormat(BaseSpaceFormat):
+    @property
+    def space(self):
+        return _VEC3_SIGNED_FLOAT32_SPACE
+
+
+class JerkSpaceFormat(BaseSpaceFormat):
+    @property
+    def space(self):
+        return _VEC3_SIGNED_FLOAT32_SPACE
+
+
+class ConfigurableSpaceFormat(BaseSpaceFormat):
+    def __init__(self, agent_interface: AgentInterface) -> None:
+        self._agent_interface = agent_interface
+
+    def active(self, agent_interface: AgentInterface) -> bool:
+        raise NotImplementedError()
+
+    @property
+    def name(self):
+        raise NotImplementedError()
+
+    @property
+    def space(self):
+        raise NotImplementedError()
+
+
+class Image8BSpaceFormat(ConfigurableSpaceFormat):
+    def __init__(
+        self, agent_interface: AgentInterface, dimensions, colors: int
+    ) -> None:
+        self._dimensions = dimensions
+        self._colors = colors
+        super().__init__(agent_interface)
+
+    @property
+    def space(self):
+        return gym.spaces.Box(
+            low=0,
+            high=255,
+            shape=(self._dimensions.height, self._dimensions.width, self._colors),
+            dtype=np.uint8,
+        )
+
+
+class DictSpaceFormat(ConfigurableSpaceFormat):
+    def __init__(self, agent_interface, spaces: List[BaseSpaceFormat]) -> None:
+        self._spaces = [space for space in spaces if space.active(agent_interface)]
+        super().__init__(agent_interface)
+
+    def format(self, obs: Observation):
+        return {s.name: s.format(obs) for s in self._spaces}
+
+    def active(self, agent_interface: AgentInterface) -> bool:
+        raise NotImplementedError()
+
+    @property
+    def name(self):
+        raise NotImplementedError()
+
+    @property
+    def space(self):
+        return gym.spaces.Dict({s.name: s.space for s in self._spaces})
```

Comment:
Is the space composition executed multiple times when called repeatedly and, if yes, can it be prestored?

Reply:
Duplicate. https://github.com/huawei-noah/SMARTS/pull/1793#discussion_r1081643799